<p style="text-align: left;"> Uğur Erdem Seyfi <br> 21801744 </p>


# <p style="text-align: center;"> MATH420 HW03 </p>


# 4.2

**Question:**

Samantha uses the RSA signature scheme with public modulus $N = 1562501$ and public verification exponent $e = 87953$. Adam claims that Samantha has signed each of the documents:

$$ D = 119812, D' = 161153, D''= 586036 $$

and that the associated signatures are

$$ S = 876453, S' = 870099, S''= 602754 $$

Which of these are valid signatures?

**Solution:**

$ S^e \equiv 876453^{87953} \equiv 772481 \text{ (mod 1562501) }$ since $ S^e = 772481 \neq 119812 = D$ S is NOT a valid signature.

$ S'^e \equiv 870099^{87953} \equiv 161153 \text{ (mod 1562501) }$ since $ S'^e = 161153 = D' $ S' is a valid signature.

$ S''^e \equiv 602754^{87953} \equiv586036 \text{ (mod 1562501) }$ since $ S''^e = 586036 = D'' $ S'' is a valid signature.

So S' and S'' are valid signatures.

# 4.3

**Question:**

Samantha uses the RSA signature scheme with public modulus and public verification exponent $ N = 27212325191$ and $e = 22824469379$.

Use whatever method you want to factor N, and then forge Samantha’s signature on the document $D = 12910258780$.

**Solution:**

Let's use pollard's method for finding a factor $ p$ of $ N = 27212325191$. 


In [9]:
# input: Integer N to be factored
def pollardsFactorization(N):
    # 1. Set a = 2
    a = 2
    # 2. Loop j = 2,3,4.. up to a specified bound
    j = 2
    while(True):
        # 3. set a = a^j mod N
        a = a^j % N
        # 4. compute gcd(a-1, N)
        d = gcd(a - 1, N)
        # if 1 < d < N, then success
        if(1 < d < N):
            return d
        
        j = j + 1

        
N = 27212325191
fact = pollardsFactorization(N)
print("First factor of N = 27212325191 is ", fact)
print("Second factor of N = 27212325191 is ", N / fact)

First factor of N = 27212325191 is  128311
Second factor of N = 27212325191 is  212081


Thus we have $ N = pq $ where we have primes $p = 128311$ and $q = 212081$. Since we should have $ ed = 1 \text{ (mod (p-1)(q-1) )}$, we have $ 22824469379d = 1 \text{ (mod 27211984800)} $, by solving this we obtain $ d = 18408628619$.

So by signing D with d, we get $ D^d \equiv 12910258780^{18408628619} \equiv 22054770669 \text{ (mod 27212325191)} $ thus we have:

$$ S = 22054770669 $$

# 4.4

**Question:**

**Solution:**



# 4.6

**Question:**

Samantha uses the Elgamal signature scheme with prime p = 6961 and primitive root g = 437. Her public verification key is A = 4250. Adam claims that Samantha has signed each of the documents

$$ D = 1521, D' = 1837, D'' = 1614 $$

and that the associated signatures are:

$$ (S_1, S_2) = (4129, 5575), (S'_1, S'_2) = (3145, 1871), (S''_1 , S''_2 ) = (2709, 2994) $$

Which of these are valid signatures?

**Solution:**

For $(S_1, S_2) = (4129, 5575)$, we have:

$$ v^{S_1} S_1^{S_2} \equiv (4250^{4129}). 4129^{5575} \equiv 231 \text{ (mod 6961)}$$

and we also have:

$$ g^D \equiv 437^{1521} \equiv 231 \text{ (mod 6961)}$$

thus we have $ v^{S_1} S_1^{S_2} \equiv g^D \text{ (mod 6961)} $ hence (S1, S2) = (4129, 5575) is valid.


For $(S'_1, S'_2) = (3145, 1871)$, we have:

$$ v^{S'_1} S_1^{S'_2} \equiv (4250^{3145}) 3145^{1871} \equiv 6208 \text{ (mod 6961)}$$

and we also have:

$$ g^{D'} \equiv 437^{1837} \equiv 2081 \text{ (mod 6961)}$$

thus we have $ v^{S'_1} S_1^{S'_2} \not\equiv g^{D'} \text{ (mod 6961)} $ hence (S'_1, S'_2) = (4129, 5575) is NOT valid.

For $(S''_1 , S''_2 ) = (2709, 2994)$, we have:

$$ v^{S''_1} S_1^{S''_2} \equiv (4250^{2709}).2709^{2994} \equiv 2243 \text{ (mod 6961)}$$

and we also have:

$$ g^{D''} \equiv 437^{1614} \equiv 2243 \text{ (mod 6961)}$$

thus we have $ v^{S''_1} S_1^{S''_2} \not\equiv g^{D''} \text{ (mod 6961)} $ hence (S''_1, S''_2) = (2709, 2994) is valid.

# 4.8 

**Question:**

Suppose that Samantha is using the Elgamal signature scheme and that she is careless and uses the same random element $k$ to sign two documents $D$ and $D'$

**(a)** Explain how Eve can tell at a glance whether Samantha has made this mistake.

**(b)** If the signature on D is $(S_1, S_2)$ and the signature on D' is $(S'_1, S'_2)$, explain
how Eve can recover a, Samantha’s private signing key.

**(c)** Apply your method from (b) to the following example and recover Samantha’s signing key a, where Samantha is using the prime p = 348149, base g = 113459, and verification key A = 185149.

$$ D = 153405, S_1 = 208913, S_2 = 209176 $$

$$ D'= 127561, S'_1 = 208913, S'_2 = 217800 $$

**Solution:**

**(a)**

Since $S_1 = S'_1$, this means that we have $g^{e} = g^{e'}$. Eve can use this fact to realize Samantha used the same ephereal key. 

**(b)**

Now by discrete logarithms, we have the verification conditions:

$$ S_1 log_g(v) + S_2 log(S_1) \equiv D \text{ (mod p - 1)} $$
$$ S'_1 log_g(v) + S'_2 log(S'_1) \equiv D' \text{ (mod p - 1)} $$



**(c)**


# 6.5(d)

**Question:** 

For each of the following elliptic curves $E$ and finite fields $\mathbb{F}_p$, make a list of the set of points $E(\mathbb{F}_p)$

$$ E : Y^2 = X^3 + 9X + 5 \text{ over } \mathbb{F}_{11}$$

**Solution:**


In the algorithm belove, we first find the generator of the EllipticCurve and then using that generator generate the points that is on our EllipticCurve. 

So our points are: $ \lbrace(2, 8), (0, 7), (1, 9), (9 , 10 , 1), (3, 9), (7, 9), (6, 0), (7, 2), (3, 2), (9, 1), (1, 2), (0, 4), (2, 3), \theta \rbrace$




In [21]:
R = Zmod(11)
# lets construct an elliptic curve over R
E = EllipticCurve(R, [9, 5])
print(E)
P = E.gens()[0] # you can calculate the generator of this curve like this
print("Generator of the Ring E(F_p) is: ", P)

points = [P]
n = 2
while( n*P != P):
    points.append( n*P)
    n = n+1

print("Generated points are: ", points)

Elliptic Curve defined by y^2 = x^3 + 9*x + 5 over Ring of integers modulo 11
Generator of the Ring E(F_p) is:  (2 : 8 : 1)
Generated points are:  [(2 : 8 : 1), (0 : 7 : 1), (1 : 9 : 1), (9 : 10 : 1), (3 : 9 : 1), (7 : 9 : 1), (6 : 0 : 1), (7 : 2 : 1), (3 : 2 : 1), (9 : 1 : 1), (1 : 2 : 1), (0 : 4 : 1), (2 : 3 : 1), (0 : 1 : 0)]



# 6.6 (c)

**Question:**

Make an addition table for $E$ over $\mathbb{F}_p$,

$$ Y^2 = X^3 + 2X + 5 \text{ over } \mathbb{F}_11 $$

**Solution:**

We have $E(\mathbb{F}_{11}) = \lbrace \theta, (0, 4), (0, 7), (3, 4), (3, 7), (4, 0), (8, 4), (8, 7), (9, 2), (9, 9) \rbrace$.

And thus our multiplication table is like the following:

\begin{array}{|c|c|c|c|c|c|c|c|c|c|}
    \hline
   {}     & \theta & (0, 4) & (0, 7) & (3, 4) & (3, 7) &(4, 0) & (8, 4) & (8, 7) & (9, 2) & (9, 9) \\ \hline
   \theta & \theta & (0, 4) & (0, 7) & (3, 4) & (3, 7) & (4, 0) & (8, 4) & (8, 7) & (9, 2) & (9, 9) \\ \hline
   (0, 4) & (0, 4) & (9, 2) & \theta & (8, 7) & (9, 9) & (8, 4)&  (3, 7) & (4, 0) & (3, 4) & (0, 7) \\ \hline
    (0, 7) & (0, 7) & \theta & (9, 9) & (9, 2) & (8, 4) & (8, 7) & (4, 0) & (3, 4) & (0, 4) & (3, 7) \\ \hline
    (3, 4) & (3, 4) & (8, 7) & (9, 2) & (8, 4) & \theta & (9, 9) & (0, 7) & (3, 7) & (4, 0) & (0, 4) \\ \hline
    (3, 7) & (3, 7) & (9, 9) & (8, 4) & \theta & (8, 7) & (9, 2) & (3, 4) & (0, 4) & (0, 7) & (4, 0) \\ \hline
    (4, 0) & (4, 0) & (8, 4) & (8, 7) & (9, 9) & (9, 2) & \theta & (0, 4) & (0, 7) & (3, 7) & (3, 4) \\ \hline
    (8, 4) & (8, 4) & (3, 7) & (4, 0) & (0, 7) & (3, 4) & (0, 4) & (9, 2) & \theta & (9, 9) & (8, 7) \\ \hline
    (8, 7) & (8, 7) & (4, 0) & (3, 4) & (3, 7) & (0, 4) & (0, 7) & \theta & (9, 9) & (8, 4) & (9, 2) \\ \hline
    (9, 2) & (9, 2) & (3, 4) & (0, 4) & (4, 0) & (0, 7) & (3, 7) & (9, 9) & (8, 4) & (8, 7) & \theta \\ \hline
    (9, 9) & (9, 9) & (0, 7) & (3, 7) & (0, 4) & (4, 0) & (3, 4) & (8, 7) & (9, 2) & \theta & (8, 4) \\ \hline
\end{array}


# 6.8

**Question:**

Let E be the elliptic curve $ E : y^2 = x^3 + x + 1 $ and let $ P = (4, 2) $ and $ Q = (0, 1) $ be points on $E$ modulo $5$. Solve the elliptic curve discrete logarithm problem for $P$ and $Q$, that is, find a positive integer $n$ such that
$Q = nP$

**Solution:**

By using discrete logarithm for P and Q in the belove code, we obtain $ n = 5$

In [14]:
(0, 7) & (0, 7) O (9, 9) (9, 2) (8, 4) (8, 7) (4, 0) (3, 4) (0, 4) (3, 7) \\ \hline
(3, 4) & (3, 4) (8, 7) (9, 2) (8, 4) O (9, 9) (0, 7) (3, 7) (4, 0) (0, 4) \\ \hline
(3, 7) & (3, 7) (9, 9) (8, 4) O (8, 7) (9, 2) (3, 4) (0, 4) (0, 7) (4, 0) \\ \hline
(4, 0) & (4, 0) (8, 4) (8, 7) (9, 9) (9, 2) O (0, 4) (0, 7) (3, 7) (3, 4) \\ \hline
(8, 4) & (8, 4) (3, 7) (4, 0) (0, 7) (3, 4) (0, 4) (9, 2) O (9, 9) (8, 7) \\ \hline
(8, 7) & (8, 7) (4, 0) (3, 4) (3, 7) (0, 4) (0, 7) O (9, 9) (8, 4) (9, 2) \\ \hline
(9, 2) & (9, 2) (3, 4) (0, 4) (4, 0) (0, 7) (3, 7) (9, 9) (8, 4) (8, 7) O \\ \hline
(9, 9) & (9, 9) (0, 7) (3, 7) (0, 4) (4, 0) (3, 4) (8, 7) (9, 2) O (8, 4) \\ \hlineR = Zmod(5)
# lets construct an elliptic curve over R
E = EllipticCurve(R, [1, 1])
print(E)
# P = E.gens()[0] # you can calculate the generator of this curve like this
# print(P )

P = E(4, 2)
print("We have P:", P)
Q = E(0,1)
print("We have Q:", Q)

xbar = P.discrete_log(Q) 
print("for Q = nP, we have n = ", xbar)

Elliptic Curve defined by y^2 = x^3 + x + 1 over Ring of integers modulo 5
We have P: (4 : 2 : 1)
We have Q: (0 : 1 : 1)
for Q = nP, we have n =  5



# 6.14

**Question:**

Alice and Bob agree to use elliptic Diffie–Hellman key exchange with the prime, elliptic curve, and point
$ p = 2671$, $E: Y^2 = X^3 + 171X + 853$, $P= (1980, 431) \in E(F_{2671})$.

**(a)** Alice sends Bob the point $Q_A = (2110, 543)$. Bob decides to use the secret
multiplier $n_B = 1943$. What point should Bob send to Alice?

**(b)** What is their secret shared value?

**(c)** How difficult is it for Eve to figure out Alice’s secret multiplier $n_A$? If you know
how to program, use a computer to find $n_A$.

**(d)** Alice and Bob decide to exchange a new piece of secret information using the same prime, curve, and point. This time Alice sends Bob only the x-coordinate $x_A = 2$ of her point $Q_A$. Bob decides to use the secret multiplier $n_B = 875$.
What single number modulo p should Bob send to Alice, and what is their secret shared value?

**Solution:**

Calculations (codes) that are related to the solutions belove are given at the end of this solution part.

**(a)**:

Bob should send $n_BP$ modulo p to the Alice, in other words, Bob should send $ n_B P = 1943 (1980, 431) = (1432, 667)$ in $E(F_{2671})$.

**(b)**:

Their secret shared value is $ n_B Q_A = 1943.(2110, 543) = (2424, 911) $ in $E(F_{2671})$.

**(c)**:

It is as hard as solving the discrete logarithm problem for Elliptic Curve Cryptography. 

By our computation, we find $n_A = 726$.


**(d)**:

Bob should send $ n_B P = 875 (1980, 431) = (161, 2040)$ in $E(F_{2671})$.

In order to determine their secret shared value, we should be able to find $Q_A$ by $x_A$, plugging $x_A$ to the elliptic curve in $E(F_{2671})$, we obtain: $ y^2 = 2^3 + 171 * 2 + 853 = 1203 $ modulo 2671, now by proposition 2.26 in the book since $ 2671 \equiv 3 \text{ (mod 4)}$ and since  $ y^2 = 2^3 + 171 * 2 + 853 = 1203 $ modulo 2671 should have a solution, we have the particular solution $y_B = \pm 1203^{(2671+1)/4} $ modulo 2671. Thus we have $y_B = \pm 2575 $. Thus $$ Q_A = \pm (2, 2575) $$ and hence their secret shared value K is:

$$ K = n_B Q_A = \pm 875 (2, 2575) = \pm(1708, 1419) \text{ (mod 2671) }$$

In [35]:
R = Zmod(2671)
E = EllipticCurve(R, [171, 853])
print(E)

# part a
print("a)")
P = E(1980, 431)
Q_B = 1943 * P
print("Bob sends ", Q_B, " to ALice.")

# part b
print("b)")
Q_A = E(2110, 543)
K = 1943 * Q_A
print("Their secretly shared key is", K)

# part c
print("c)")
n_a = P.discrete_log( Q_A)
print("n_a is", n_a)

# part d
print("d)")
Q_B = 875 * P
print("Bob sends ", Q_B, " to ALice.")
print("y^2 =", (2^3 + 171 * 2 + 853) % 2671, " modulo 2671")
print("y_B = (-+)", 1203^((2671+1)/4) % 2671)
print("Their secret shared key is: (-+)", 875 * E(2, 2575))

Elliptic Curve defined by y^2 = x^3 + 171*x + 853 over Ring of integers modulo 2671
a)
Bob sends  (1432 : 667 : 1)  to ALice.
b)
Their secretly shared key is (2424 : 911 : 1)
c)
n_a is 726
d)
Bob sends  (161 : 2040 : 1)  to ALice.
y^2 = 1203  modulo 2671
y_B = (-+) 2575
Their secret shared key is: (-+) (1708 : 1419 : 1)


# 6.17

--

# 6.20

--

# 6.21(d)

Use the elliptic curve factorization algorithm to factor each of the numbers $N$ using the given elliptic curve $E$ and point $P$.

$$ N = 28102844557, E: Y^2 = X^3 + 18X − 453, P= (7, 4) $$